### ISSUE TITLE
Unconfirmed transactions created with `sendrawtx_MP` RPC don't show up as Omni Transaction

### ISSUE THREAD

**Issue by msgilligan** [OmniLayer/omnicore#299]
I created some transactions with the `sendrawtx_MP` RPC (in this case createProperty / Tx50) and they showed up in **Transactions -> Bitcoin** as a confusing/mysterious transaction, but did not show up in **Transactions -> Omni Layer**.

**Comment by dexX7** [comment]
> confusing/mysterious transaction

What's that? Do you mean conflicting by any chance? If so, this is a transaction, which is not part of the active chain (e.g. burried due to block reorganizations).

Is this reproducible? Was this a result of the regtests?

Maybe you could paste the result from `getrawtransaction <txid>` here.

**Comment by msgilligan** [comment]
When saying "confusing/mysterious" I was putting myself in the mindset of an end user. What I mean is that it showed up as a small value Bitcoin transaction when it was really an Omni transaction. I don't remember if the `Address` field was labelled `(Omni transaction)` or not.

This was not a result of RegTest. I was creating a new property (SafeExchangeCoin) on Main Net and wrote a Groovy script to do it.

Txid: `0b61c14741228a50565769ab665998d96f2da0de6b1f19ec59fc11101092cbc5`

**Comment by msgilligan** [comment]
Raw:

    getrawtransaction 0b61c14741228a50565769ab665998d96f2da0de6b1f19ec59fc11101092cbc5

    01000000012c2cf122cf1bc8924077ca98496da2db4ffee3d213161a9781943ac9d196a62e000000006b483045022100a6a1c00ee3fc6066cfce34c289e752e0ddfca22c5e18a9b946f990189d17824d022061a0a8accf86ed1bde781e9fd83cfe7dbd162c881698bcef9282283490a8ab900121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914effffffff055a0f000000000000695121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e21029513d6e1dc89455622c021a2abf0fd6a05a13b01db762e6c66e6586352e7c44e2102296f23e9797cea71f476156ccf243de01590a0d1853ca014f00c7b5843297efa53ae5a0f000000000000695121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e2102bf961b3065f97635272d5db7892c52d16708578fb3338210141f341dd663a8d121023d7efd01b65b561e66ecaf3f1fe29c116f9f8dcd60e5e1940966bdf34384df9353ae5c0d000000000000475121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e2102c099cdaf87465767dd51231d6602f2825f0c4c84362ae3bcb397f30cd51391c852ae46d81d00000000001976a914ca6e02fd4241938f7c9a00beaa3670d57dcb17c588acaa0a0000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac00000000


JSON:

    gettransaction 0b61c14741228a50565769ab665998d96f2da0de6b1f19ec59fc11101092cbc5


```    
{
"amount" : -0.00014010,
"fee" : -0.00010000,
"confirmations" : 1717,
"blockhash" : "0000000000000000076f72fd3d0049f13b1902661531eee5850593d10f5a3fd9",
"blockindex" : 89,
"blocktime" : 1448159454,
"txid" : "0b61c14741228a50565769ab665998d96f2da0de6b1f19ec59fc11101092cbc5",
"walletconflicts" : [
],
"time" : 1448159362,
"timereceived" : 1448159362,
"details" : [
{
"account" : "",
"category" : "send",
"amount" : -0.00003930,
"vout" : 0,
"fee" : -0.00010000
},
{
"account" : "",
"category" : "send",
"amount" : -0.00003930,
"vout" : 1,
"fee" : -0.00010000
},
{
"account" : "",
"category" : "send",
"amount" : -0.00003420,
"vout" : 2,
"fee" : -0.00010000
},
{
"account" : "",
"address" : "1KTMHViPPKM3Y8KnJqBCnLkAsSZPuxaokp",
"category" : "send",
"amount" : -0.01955910,
"vout" : 3,
"fee" : -0.00010000
},
{
"account" : "",
"address" : "1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P",
"category" : "send",
"amount" : -0.00002730,
"vout" : 4,
"fee" : -0.00010000
},
{
"account" : "Safe Exchange Issuing Address",
"address" : "1KTMHViPPKM3Y8KnJqBCnLkAsSZPuxaokp",
"category" : "receive",
"amount" : 0.01955910,
"vout" : 3
}
],
"hex" : "01000000012c2cf122cf1bc8924077ca98496da2db4ffee3d213161a9781943ac9d196a62e000000006b483045022100a6a1c00ee3fc6066cfce34c289e752e0ddfca22c5e18a9b946f990189d17824d022061a0a8accf86ed1bde781e9fd83cfe7dbd162c881698bcef9282283490a8ab900121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914effffffff055a0f000000000000695121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e21029513d6e1dc89455622c021a2abf0fd6a05a13b01db762e6c66e6586352e7c44e2102296f23e9797cea71f476156ccf243de01590a0d1853ca014f00c7b5843297efa53ae5a0f000000000000695121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e2102bf961b3065f97635272d5db7892c52d16708578fb3338210141f341dd663a8d121023d7efd01b65b561e66ecaf3f1fe29c116f9f8dcd60e5e1940966bdf34384df9353ae5c0d000000000000475121021b28aeea9d49dd9c10b80b0d3b76d151a0cdd8bf44b5203e04b5132aa5a2914e2102c099cdaf87465767dd51231d6602f2825f0c4c84362ae3bcb397f30cd51391c852ae46d81d00000000001976a914ca6e02fd4241938f7c9a00beaa3670d57dcb17c588acaa0a0000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac00000000"
}
```

**Comment by dexX7** [comment]
Aah, now I understand, thanks for the information!

Currently not all transaction types show up immediately as unconfirmed transaction, but only after the confirmation.

**Comment by dexX7** [comment]
Currently not showing up as unconfirmed:

- Send All
- DEx Accept
- SP related transactions

**Comment by dexX7** [comment]
After looking further into this: it's actually a topic @zathras-crypto and I discussed earlier.

Currently the flow is as follows:

1. user sends a transaction via GUI or RPC
2. the transaction is created and broadcasted
3. a pending transaction object is created, consisting of:
  1. txid
  2. sending address
  3. transaction type
  4. property identifier
  5. amount
  6. a flag, whether the specified amount should be subtracted from the "pending balance"
4. during the GUI update, the pending transaction objects (amongst other transactions) are processed

If I recall correctly, some transaction types don't trigger new pending transaction objects, because we haven't yet decided how they are shown in the UI (e.g.: send-all has no single property identifier attached).

**Comment by zathras-crypto** [comment]
I can definitely look into the pending objects for:

> Currently not showing up as unconfirmed:
omni_sendall
omni_senddexaccept
omni_sendissuancecrowdsale
omni_sendissuancefixed
omni_sendissuancemanaged
omni_sendgrant
omni_sendrevoke
omni_sendclosecrowdsale
omni_sendchangeissuer
omni_sendactivation
omni_sendalert

Because these use the ```omni_send*``` logic (ie they leverage the wallet builder).

The ```omni_sendrawtx``` call however (I believe that was the original issue) does not interact with the wallet builder, and does not place a transaction in the wallet (remember raw txs do not have to be wallet txs).  Transaction history is based on wallet history.

TL:DR; I'm sure I can fix up the various ```omni_send*```'s but providing visibility in the UI for raw sent transactions might be difficult, and if they didn't involve anything in the wallet might be very difficult.

Leave it with me :)

Thanks
Z

**Comment by dexX7** [comment]
To tackle this, I could imagine to add a hook to the missing RPCs, although it's still an open topic how to deal with amounts and property identifiers. In case of doubt, we could just use the ecosystem instead of a token identifier, and use zero amounts. Then on the UI side, we'd probably need to add a few more `"N/A"` paths for these transactions.

In the longer run we should probably get away from the approach of using explicitly added pending transaction objects, and instead parse transactions traditionally (`GetTransaction` -> `ParseTransaction` -> interpret), given that we can deal with unconfirmed transactions pretty good at this point. This might even be done right after sending any transaction. In an even longer timeframe this process could be triggered by the mempool, but I think we're not there yet.

**Comment by dexX7** [comment]
Over the next days I'm going to take a closer look at this, and as first step my plan is to get rid of the pending transaction objects (as far as possible), and potentially just store the hash of pending Omni transactions and then parse on the fly.

**Comment by dexX7** [comment]
Here is roughly the game plan:

**Isolation and extraction of transaction parsing related code:**

1. [overviewpage.cpp#L116-L226](https://github.com/OmniLayer/omnicore/blob/363569772e029f7fee61efa3bc6d35de1ff91394/src/qt/overviewpage.cpp#L116-L226): extract creation of OverviewCacheEntry
2. [txhistorydialog.cpp#L225-L341](https://github.com/OmniLayer/omnicore/blob/363569772e029f7fee61efa3bc6d35de1ff91394/src/qt/txhistorydialog.cpp#L225-L341): extract creation of HistoryTXObject
3. [tradehistorydialog.cpp#L259-L432](https://github.com/OmniLayer/omnicore/blob/64847caf6152954f90f4f6095ff0048673ab1df7/src/qt/tradehistorydialog.cpp#L259-L432): extract creation of TradeHistoryObject

**Related objects:**

OverviewCacheEntry:
- QString address
- QString amount
- bool valid
- bool sendToSelf
- bool outbound

HistoryTXObject:
- int blockHeight
- int blockByteOffset
- bool valid
- bool fundsMoved
- std::string txType
- std::string address
- std::string amount

TradeHistoryObject:
- int blockHeight
- bool valid
- uint32_t propertyIdForSale
- uint32_t propertyIdDesired
- int64_t amountForSale
- std::string status
- std::string info
- std::string amountOut
- std::string amountIn

The underlying code for all three is pretty similar (iterate over pending transactions, iterate over wallet transactions, do some magic and create the objects), so it seems like duplicated effort to do it three times. Combining OverviewCacheEntry and HistoryTXObject seems straight forward, because a history object is basically an extended overview entry. Not sure how to deal with trade objects though, because not all transactions are trades, but basically all transactions should show up on the overview tab and the transaction history.

**Create one representation for all**

1. Create a new class to combine OverviewCacheEntry, HistoryTXObject (and maybe TradeHistoryObject), which shares the properties of these
2. Update and merge the previously extracted methods to create the the new entry objects
3. Adjust the GUI code to use the new class

Open question: right now the objects are created on the fly during an UI update (e.g. from wallet cache, by iterating over pending transactions etc.). If the entry creation is merged and done for both or all three at the same time, then there should only be one place where it's done once.

**Detach creation of objects from GUI updates**

Instead of iterating over the wallet (...) on the fly and creating the objects, during an UI update the objects should already be available (e.g. in form of a cache), and the actual creation should be done unrelated to the UI updates. In an ideal scenario we'd somehow tap the mempool or make use of the transaction/wallet signals for transactions and their status. Alternatively we could continue to use our own signals.

Entries have stateless and stateful properties, such as the stateless transaction type, amount etc., and the stateful confirmations, which should be updated accordingly, if their status changes.

Does all this somehow make sense? The plan isn't complete, and there are probably quite a few parts missing, and I'm not yet sure, if this can be integrated nicely.

My goal right now is to end up with "one batch of entries", which is mostly detached from the GUI. If this is done, the actual UI integration is hopefully more or less straight forward.

**Comment by dexX7** [comment]
Unconfirmed transactions should now show up via `omni_gettransaction`.
